src/Storage/Repository/PgInstanceAndEventsRepository.cs

```diff
@@ -0,0 +1,83 @@
+{
+    private readonly ILogger<PgInstanceAndEventsRepository> _logger;
+    private readonly NpgsqlDataSource _dataSource;
+    private readonly TelemetryClient _telemetryClient;
+    private readonly IInstanceRepository _instanceRepository;
+
+    private readonly string _insertInstanceEventsSql = "call storage.insertinstanceevents($1)";
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="PgInstanceAndEventsRepository"/> class.
+    /// </summary>
+    /// <param name="logger">The logger to use when writing to logs.</param>
+    /// <param name="dataSource">The npgsql data source.</param>
+    /// <param name="instanceRepository">Instance repo</param>
+    /// <param name="telemetryClient">Telemetry client</param>
+    public PgInstanceAndEventsRepository(
+        ILogger<PgInstanceAndEventsRepository> logger,
+        NpgsqlDataSource dataSource,
+        IInstanceRepository instanceRepository,
+        TelemetryClient telemetryClient = null)
+    {
+        _logger = logger;
+        _dataSource = dataSource;
+        _instanceRepository = instanceRepository;
+        _telemetryClient = telemetryClient;
+    }
+
+    /// <inheritdoc/>
+    public async Task<Instance> Update(Instance instance, List<string> updateProperties, List<InstanceEvent> events)
+    {
+        if (events.Count == 0)
+        {
+            return await _instanceRepository.Update(instance, updateProperties);
+        }
+        
+        // Remove last decimal digit to make postgres TIMESTAMPTZ equal to json serialized DateTime
+        instance.LastChanged = instance.LastChanged != null ? new DateTime((((DateTime)instance.LastChanged).Ticks / 10) * 10, DateTimeKind.Utc) : null;
+        List<DataElement> dataElements = instance.Data;
+
+        PgInstanceRepository.ToInternal(instance);
+        instance.Data = null;
+        await using NpgsqlBatch batch = _dataSource.CreateBatch();
+
+        NpgsqlBatchCommand updateCommand = new(PgInstanceRepository.UpdateSql);
+        PgInstanceRepository.BuildUpdateCommand(instance, updateProperties, updateCommand.Parameters);
+        batch.BatchCommands.Add(updateCommand);
+
+        NpgsqlBatchCommand insertEventsComand = new(_insertInstanceEventsSql);
+        var records = events.Select(e => new InstanceEventRecord(new Guid(e.InstanceId.Split('/').Last()), e.Id, e)).ToArray();
+        insertEventsComand.Parameters.AddWithValue(NpgsqlDbType.Jsonb, records);
+        batch.BatchCommands.Add(insertEventsComand);
+
+        await using NpgsqlDataReader reader = await batch.ExecuteReaderAsync();
+
+        if (await reader.ReadAsync())
+        {
+            instance = await reader.GetFieldValueAsync<Instance>("updatedInstance");
+        }
+
+        instance.Data = dataElements; // TODO: requery instead?
+
+        return PgInstanceRepository.ToExternal(instance);
+    }
+
+    private sealed record InstanceEventRecord(Guid Instance, Guid? AlternateId, InstanceEvent Event);
```

Comment:
Why this new record? Why not sending List<InstanceEvent> to the procedure?

Reply:
I just made it as similar to `storage.insertInstanceevent`/`InsertInstanceEvent` as I could. I'm fine with just using `InstanceEvent`, but then I would have to do the InstanceId string splitting for the first column in psql instead right?

Reply:
Sounds good to me, done that now